### ISSUE TITLE
Bulk actions

### ISSUE THREAD

**Issue by thany** [streetcomplete/StreetComplete#2996]
**Use case**

In my neighbourhood there a LOT of small buildings that I would love to start classifying as "Shed". But there are so many of them, that it's hopeless to do them one by one. I mean, that would take loads of time, wouldn't it. I think my battery wouldn't even last that long 😀
The same goes for the 100-odd houses packed together, just in my own neightbourhood (so who knows what lurks outside of it 👻) most of which have identical flooring, which is another quest on virtually all of them.

**Proposed Solution**

Maybe select-and-hold, which I think is a pretty common pattern on Android, and then just *tap*, *tap*, *tap*, *tap* (and 100 more) to select identical quests that will have identical answers, and then submit the whole lot in one go. We do have to make sure that mistapping a quest doesn't discard the selection.

**Comment by westnordost** [comment]
For such bulk edits, it is more efficient if you used the in-browser editor on www.openstreetmap.org. If you press shift, you can select multiple buildings and then tag them all at once. At least which buildings are detached houses in a suburbian district should be well visible from aerial imagery.

StreetComplete should only be used while actually on-site. If it was possible to do a bulk action, that would defeat the purpose.

The exact same feature had been requested before a couple of times, but right now I am too lazy to search for the other issues.

Also note, that I am considering turning off those extremely spammy quests (building type, building levels and roof) by default, so that users don't get bored so much of that.

**Comment by mnalis** [comment]
> Also note, that I am considering turning off those extremely spammy quests (building type, building levels and roof) by default, so that users don't get bored so much of that.

Will turning off that spammy `building type` quest also turn off very useful `house address` quests, or will those useful quests remain available (as suggested in https://github.com/streetcomplete/StreetComplete/issues/2464)?

**Comment by andrewharvey** [comment]
> At least which buildings are detached houses in a suburbian district should be well visible from aerial imagery.

One thing I've learnt from doing StreetComplete is aerials are no match for on the ground. Many houses which you'd think were detached from the imagery are actually semi-detached.

I agree with @westnordost we should not support this kind of bulk action in StreetComplete, you'll potentially make errors that reviewing each building individually would avoid.

> Also note, that I am considering turning off those extremely spammy quests (building type, building levels and roof) by default, so that users don't get bored so much of that.

Please if you are going to do that, just disable by default but allow users to turn the quest on. In my opinion StreetComplete is way too editorial and it should support way more quests that were rejected as too spammy.

**Comment by matkoniecz** [comment]
> Please if you are going to do that, just disable by default but allow users to turn the quest on. In my opinion StreetComplete is way too editorial and it should support way more quests that were rejected as too spammy.

Note that it proposes solely change in default settings - enabling this quests would be still possible.

**Comment by thany** [comment]
> For such bulk edits, it is more efficient if you used the in-browser editor on [www.openstreetmap.org](http://www.openstreetmap.org). If you press shift, you can select multiple buildings and then tag them all at once. At least which buildings are detached houses in a suburbian district should be well visible from aerial imagery.

I've tried, but the editor they provide doesn't really agree with me. Bulk actions are not a great experience. It seems more tailored towards detailed edits, and less toward bulk actions. Sure, possible, but still prohibitively tedious, especially when a selection of 200 sheds is cancelled just because I accidentally click without holding shift 🤨
Anyway, not your problem, but now maybe you have some idea where I'm coming from.

> StreetComplete should only be used while actually on-site. If it was possible to do a bulk action, that would defeat the purpose.

Absolutely agree, but in my scenario it's about my own neighbourhood. I've lived here for 14 years, so I have a fairly good idea what houses are like around here. So there's a tradeoff to be made: one the one hand, you don't want to encourage bulk actions over a huge area, but on the other hand, even when on-site, if you're standing in front of a row of 20 houses that you can see all at once in your perpheral view, it does make sense to answer a question for all of those house in one go.

> The exact same feature had been requested before (and rejected) a couple of times, but right now I am too lazy to search for the other issues.

I believe you 😉
But that only enforces the demand that exists for a feature like this, imo.

> Also note, that I am considering turning off those extremely spammy quests (building type, building levels and roof) by default, so that users don't get bored so much of that.

That's another option. But then, those question won't get answered, unless someone really wants to bore him/herself with them - still without the possibility to even do "small bulk" answers.

**Comment by mnalis** [comment]
> Sure, possible, but still prohibitively tedious, especially when a selection of 200 sheds is cancelled just because I accidentally click without holding shift 

@thany that would depend on the editor used. In [JOSM](https://wiki.openstreetmap.org/wiki/JOSM) for example, it is trivial (and much less error prone) to select all undetailed buildings in the selection (made as bbox or via lasso select or whatever), and then only filter out those marked with `building=yes` to do an action (like changing to `building=house` or whatever) on them just by using `ctrl-f`. (I don't use iD much so can't help you there)

> if you're standing in front of a row of 20 houses that you can see all at once in your perpheral view, it does make sense to answer a question for all of those houses in one go.

Note that StreeteComplete does **save the most-recently-used** building type for that quest, so if next building is the same type as previous one, it is much easier to mark it as same building (then to go search for another building type). But yes, multi-select might be even easier (or it might not, see your own experience with desktop OSM editors - on mobile phone it might turn out **even worse** and more complicated, due to smallness of screen and lack of external modifiers like shift key) -- so current situation is not all that bad (and given multi-select is not going to happen, it is rather hypothetical anyway :smile: )

> That's another option. But then, those question won't get answered, unless someone really wants to bore him/herself with them - still without the possibility to even do "small bulk" answers.

This part is discussed around https://github.com/streetcomplete/StreetComplete/issues/2944#issuecomment-867530414, so feel free to look up; general idea is to (semi-?)automatically reenable them later when more important quest are cleared, or newbie becomes more stable user etc.

**Comment by eginhard** [comment]
> > Sure, possible, but still prohibitively tedious, especially when a selection of 200 sheds is cancelled just because I accidentally click without holding shift
> 
> @thany that would depend on the editor used. In [JOSM](https://wiki.openstreetmap.org/wiki/JOSM) for example, it is trivial (and much less error prone) to select all undetailed buildings in the selection (made as bbox or via lasso select or whatever), and then only filter out those marked with `building=yes` to do an action (like changing to `building=house` or whatever) on them just by using `ctrl-f`. (I don't use iD much so can't help you there)

@thany In JOSM you also have the possibility to undo a (de)selection with Ctrl+Shift+Z, i.e. to go back to whatever your previous selection was.

**Comment by thany** [comment]
@eginhard 
> In JOSM you also have the possibility to undo a (de)selection with Ctrl+Shift+Z, i.e. to go back to whatever your previous selection was.

Nope. Doesn't work. Checked the menus as well. On top of that, I can't even reliable *make* a selection in JOSM, half the time it grabs a node, even when it's clearly in lasso selection mode. Not a very well made program, iyam. Technically sound, but UI leaves some to be expected.

Anyway, that's getting offtopic. I do still think SC can be made easier for multiple actions. If I'm surrounded by 20 sheds, I would love to be able to answer "building type" questions for all of them in one go. That to me just makes a lot of sense.

**Comment by mnalis** [comment]
> Nope. Doesn't work. Checked the menus as well. On top of that, I can't even reliably _make_ a selection in JOSM, half the time it grabs a node, even when it's clearly in lasso selection mode. Not a very well made program, iyam. Technically sound, but UI leaves some to be expected.

well, `Ctrl+Shift+Z,` does work for me nicely (even if it is not in the menus) at least in JOSM `17919` under Debian GNU/Linux with icewm, although I prefer find_in_selection method I outlined before. And both lasso and bbox mode allow single-click selects too, that is by design. But as you say, this is not JOSM discussion forum :smile:. It does by chance illustrate that what is preferred UI for some people does not fit well with other people, which is quite related; see below. 

> Anyway, that's getting offtopic. I do still think SC can be made easier for multiple actions. If I'm surrounded by 20 sheds, I would love to be able to answer "building type" questions for all of them in one go. That to me just makes a lot of sense.

Can you outline in detail how exactly do you envision usable mobile UI (user interface) for this to work? I've tried to think of several ways, and all of them were **much worse in usability** then what JOSM does (and some people are not even happy with that!) Main problems seem to be lack of modifier keys, touch interface, and small screen. 

So if you try to **explain in detail from start to end for typical multiple-quest-answer** exactly where you would click, what gesture you would do, what screen would popup, etc  (preferably with screen mockups if it conveys the idea better than words), and if that idea turns out to be fast and simple to use and does not break other things (and is possible to implement under current tangrem-ES framework, of course) then perhaps someone might be tempted to try to implement it.

**Comment by eginhard** [comment]
> Nope. Doesn't work. Checked the menus as well. On top of that, I can't even reliably _make_ a selection in JOSM, half the time it grabs a node, even when it's clearly in lasso selection mode. Not a very well made program, iyam. Technically sound, but UI leaves some to be expected.

Just to finish this tangent: I see that it requires the [UtilsPlugin2](https://josm.openstreetmap.de/wiki/Help/Plugin/UtilsPlugin2) which also has a wide range of other very useful tools.

**Comment by pigsonthewing** [comment]
> For such bulk edits, it is more efficient if you used the in-browser editor on www.openstreetmap.org.

That is _not_ more efficient, nor even practicable, on a small-screen device like a phone, while walking, or in a parked car.

> StreetComplete should only be used while actually on-site. If it was possible to do a bulk action, that would defeat the purpose.

This is a non-sequitur. 

It is possible and, in my experience, usual to make repeated edits of the same type, with the same value, while on-site.

> The exact same feature had been requested before (and rejected) a couple of times, 

The fact that this is repeatedly requested is user-feedback. Please heed it.

**Comment by matkoniecz** [comment]
> user-feedback. Please heed it.

Note that in general optimal use of user feedback is to neither ignore it nor implement all widely requested features - but something in between. Design by committee is considered as a pejorative term, for quite clear reasons, and unfiltered user feedback would result in this (and other problems).

**Comment by riQQ** [comment]
The same feature was also requested in #57.

**Comment by mnalis** [comment]
> > For such bulk edits, it is more efficient if you used the in-browser editor on [www.openstreetmap.org](http://www.openstreetmap.org).
> 
> That is _not_ more efficient, nor even practicable, on a small-screen device like a phone, while walking, or in a parked car.

On a mobile device with small screen, I'd recommend [Vespucci](https://wiki.openstreetmap.org/wiki/Vespucci) instead of iD. To activate multi-select there, simply double-tap on element you want to select (instead of single tap).

In my experience though, multi-select on small screen is often likely more trouble than it is worth, unless there is huge batch of stuff to be changed (eg, when changing only one tag on mobile phone, even when there is multi select feature, I'd likely often just end up doing them one by one - because moving small screen around, zooming in/out, selecting, undoing wrong selections, verifying only wanted stuff is selected, and only then choosing the change and applying if quite likely more work and brain load and possibility of mistake then much simpler "tag, next, tag, next, tag, next"). 
But maybe I've just not yet been exposed to perfect UI. 

@pigsonthewing  Please do try Vespucci on small screen, and see how it fares for you, and how would you change its multiselect feature, if it was up to you to implement similar feature in StreetComplete. 

I'm quite confident that constructive suggestion how to implement it (in an easy and useful way, while minimizing negative effects which go with it) would go a long way toward improving chances of something like this becoming possible. 

It does require quite some effort from proponent side, though (experimenting with other solutions like Vespucci, seeing what works nicely and what doesn't on mobile, measuring how multiselect improves user experience and performance, offering detailed instructions how you feel it should be best implemented, with mockups, covering all possible logic branches  (eg. "oops, what would happen if user clicks here instead of there?") etc.

> It is possible and, in my experience, usual to make repeated edits of the same type, with the same value, while on-site.

I do have to concede that this **does hold some truth**, though. But I too am much worried about possible mistagging with it, especially when having in mind target audience of StreetComplete. If something like this were ever to be added, it should probably be only after user enables "Expert mode" https://github.com/streetcomplete/StreetComplete/discussions/3003 ?

However, it also seems to me that this annoyance factor happens mostly with low-value quests which also happen to be spammy (like `building type`, `roof shape`, `building levels`). My solution to avoid such annoyances is to disable such low-value quests unless I'm really bored and the rest of the map is solved - something that quests presets in `v34` have made quite easy.  Perhaps such "solution" could work for you too.
My main problem that remains there is that currently some useful quests (like `house number`) depend on some low-value quests (`building type`), but hopefully that issue might be reevaluated - see https://github.com/streetcomplete/StreetComplete/issues/2464

> > The exact same feature had been requested before (and rejected) a couple of times,
> 
> The fact that this is repeatedly requested is user-feedback. Please heed it.

Do you a suggestion how would that UI look and work _exactly_, and how it be keep simple and not confusing (eg. current user UI should remain the same, and there should not be an easy way for user to enter that mode by mistake)? Preventing user from selecting stuff outside of their field-of-view should also be wanted feature.

**Comment by smichel17** [comment]
> The fact that this is repeatedly requested is user-feedback

I wholeheartedly agree that solving repeated quests with the same set of answers, especially for apartment buildings, is tedious. I also agree that the multi-select/bulk-actions is the most *obvious* solution; in fact, I once [proposed it myself](https://github.com/streetcomplete/StreetComplete/issues/826). However, after giving it more though, I don't think it's the best solution to the problem.

Specifically, the problem is that it takes too many taps and too many context switches (waiting for forms/keyboard to open/close) to answer *certain* quests. I proposed an alternative solution to this problem in #1735, which I think is much better (but also a bunch of work to implement).

**Comment by thany** [comment]
I found that a feature for bulk action would come in really handy on big intersections in big North American-style cities with wide stroads tearing through the city. Each intersection has at least 8 times the quests for wheelchairs, blind folks, traffic sounds, etc. These intersections are oftenly identical on every side, making it repetitive work to tag just one intersection.

One intersection, as a result, I found can consist of 64 quests easily. While fun to do a few one by one, this kind of repetitive tagging gets old very quickly...

**Comment by e-t-l** [comment]
Regarding the original request in #57, which you rejected 6 years ago due to the infancy of this project, the point of being able to multi-select _road type_ quests in particular would be extremely useful. For example, I am 100% confident that I have walked on every road within a 1-mile radius of my home, and every single one of them has the same road type. With multi-select, I could add that information in 20 seconds. Selecting quests individually, it could take over 10 minutes - which means, as other users have pointed out, that I likely just won't do it at all.

**Comment by Helium314** [comment]
You have other tools that allow you to do that. If you know the answer anyway and don't want to go there, better use a full editor like iD instead of a tool for on-site survey.

**Comment by Cj-Malone** [comment]
I don't think I agree with this, but Apples multiple drag drop could probably inspire the implementation.

Long press one thing, while holding it select others. (It's a 2 hand operation)

**Comment by mnalis** [comment]
> That might work on Apple devices, but on Android the convention is to long-press on one thing, which then activates multiselection mode

Also, one should note that in StreetComplete  that long-press gesture is already used up for other purpose (add note /  track recording / share `geo:` in other app)

> Even Vespucci I would consider an absolute last resort

Usually, me too. If I need to add a POI or few, I do that from [Every Door](https://wiki.openstreetmap.org/wiki/Every_Door) (or occasionally [OsmAnd](https://wiki.openstreetmap.org/wiki/OsmAnd)). However, for some things, there is either adding copious notes and then waiting to go home to proper computer with mouse and big screen, or doing it just there in Vespucci, where Vespucci often wins a day for me. Especially if it involves ways or relations and not just simple nodes.

Vespucci does have a lots of features, and a rather steep learning curve, yes.  There is `Simple mode`, though, you can enable it in main menu (three vertical dots). But yes, its target audience are definitely not OSM newbies (like StreetComplete is). But note that just because it does not fit your use case, does not mean it that is bad.

In any case, bulk actions are IMHO definitely waaaay too advanced feature to drop on unsuspecting newbies, who are ofter overwhelmed even by single element quests. 

> For example: the map is greyed out when data hasn't been downloaded. First of all, I don't care, just download data if you need to

Perhaps in your country mobile data is free, and you only map for short time where battery life and phone overheating is never an issue. That is however not a case for everyone, and I for one would be absolutely horrified if general-purpose mobile OSM data editor (which Vespucci is) automatically downloaded any area I scrolled through! That being said, Vespucci has preference to enable and tune auto download settings. 

Anyway, this is not a [Vespucci](https://wiki.openstreetmap.org/wiki/Vespucci) support forum, so if you think such option should be enable by default, or have ideas how to make it simpler and easier to use _without sacrificing functionality_, feel free to suggest it there.  :smile:
